### DISCUSSION TITLE
Less Restrictive Names

### DISCUSSION THREAD

**Discussion by slanden** [pdml-lang/pdml-lang.github.io#16] [Ideas]
## Summary
I think PDML names are unecessarily restrictive given the default UTF-8 encoding and PDML's potential to be not just a replacement for XML, but an upgrade. In this proposal I'll discuss the reasons why I believe naming rules could be loosened a little.

## UTF-8
PDML uses UTF-8 encoding, except when naming things. That is, people don't get to benefit from having this encoding by default. At the time of this writing, PDML names must match the following RegEx: `[a-zA-Z_][a-zA-Z0-9_\.-]*`.

However, from a technical standpoint there are only a few characters that naturally don't make sense to include in a name, which I'll go over below.

### Characters to Prohibit
`[`

Why? This would cause the parser to stop parsing the name and start parsing a new node.

`]`

Why? This would cause the parser to end the node.

`(`

Why? This would cause the parser to start parsing attributes instead.

' ', '\t', '\n', '\r\n'

Why? The first whitespace character separates the node name from attributes or body, or attribute names from other attribute names. I'm not able to think of a reason for prohibiting other escapes though.

### Serving a Wider Audience

XML is used by all sorts of people even today for its ability for people to bring their own schemas; to make up their own node names. However, many people across the world don't speak or write English, or would prefer to use their native language when possible.

UTF-8 does a lot to help with this and I think this reason alone is enough to allow all of UTF-8 (excluding the characters mentioned previously). This would allow many more people to take advantage of the power of PDML to describe whatever they want to describe, using their own node names that makes sense to them.

## Starting With Digits
At first, it seems obvious that names shouldn't start with numbers. However, I don't see a technical reason to restrict it. Starting a "name", or *key* rather, with a digit can be useful in several cases.

For example,
```
[stats
  [per_month
    [2021 (books_sold=number books_bought=number)
      [- Data -]
    ]
    [2022 (books_sold=number books_bought=number)
      [- Data -]
    ]
  ]
  [processing_time (year=integer days=float)
    [2021 70]
    [2021 76.29]
  ]
]
```
Sometimes, organizing data by year or month can be useful when working with it in the application code.

Another example is a situation *similar* to when certain config files are prefixed with a "priority" number. E.g.
```
[01-init Initialize some things right away]
[50-meh These things might be less important]
```

## Allow Empty Names
[In another post](https://github.com/pdml-lang/pdml-lang.github.io/discussions/15) I described a use case for allowing the name to be empty. Empty names could be useful for minimizing useless bytes when used for data transfer.

```
[- Could be similar to a delimiter-separated-values list, but augmented with the power of PDML -]
[ 1 "Some data" 323 0 0]
[ 2 "More data" 750 6 0]
[ 3 "And so on" 993 0 2]
```

Another situation this could be handy is to allow having a sort of "null placeholder" in a dataset.

```
[per_month
  [2022 (books_sold=number books_bought=number)
    [books 1 2]
    [books 1 0]
  ]
  [2023 (books_sold=number books_bought=number)
    [books 0 0]
    []
    [books 3 1]
  ]
]
```

Maybe some part of application code needs there to be an exact amount of items, but the database returns less. It could probably be modeled a different way, but sometimes it's not worth it. For the second child of the `2023` node above, perhaps it's not that you bought and sold 0 books, it's that you have unknown or missing data or something to that sort.

## Case-Sensitive
Maybe names should be case-sensitive. I don't think allowing a node defined as `example` to be recognized as, `example`, `Example`, `eXAmplE`, `EXAMPLE` is a good thing.

## Reserved "Names"
I do think it'd be worth considering some things to reserve for future extensions to the language. For example, one day there might be a desire to use '!' to mean something, or '@', etc. Maybe a macro, or formatting thing, or who knows.

There may not be anything planned for them now, but it's worth thinking about.

## Conclusion

I think the current character set allowed for naming could be expanded to be more accessible to non-English writers, and to enable other use cases. And while there are only a handful of characters that make sense to prohibit, thinking about reserving some characters or words might be necessary before opening up to the rest of the UTF-8 character set.

**Comment by pdml-lang** [comment]
Preliminary answer (will follow up in a few days):

The reason for the current PDML naming rules was to ensure compatibility with the [XML naming rules](https://www.w3schools.com/xml/xml_elements.asp) (scroll down to chapter _XML Naming Rules_).

Being able to convert PDML to XML, and XML to PDML (already implemented) works well only if both formats apply the same naming rules. Therefore PDML currently applies the same naming rules as XML for nodes and attributes.

However, this isn't set in stone. You provided some valuable arguments for reevaluating the rules applied in PDML.

**Reply by slanden** [reply]
>will follow up in a few days

No rush, I just found a bit more information after reading your response and wanted to share.

So, I had a look at the [XML Specification](https://www.w3.org/TR/xml/), and at some point XML 1.0 was adjusted to loosen their naming restrictions as well. 

>  In particular, erratum [E09](http://www.w3.org/XML/xml-V10-4e-errata#E09) relaxes the restrictions on element and attribute names, thereby providing in XML 1.0 the major end user benefit currently achievable only by using XML 1.1. As a consequence, many possible documents which were not well-formed according to previous editions of this specification are now well-formed, and previously invalid documents using the newly-allowed name characters in, for example, ID attributes, are now valid.

The *Documents > Characters* section of the specification goes into more details on what's allowed in names. The linked version of the specification was published in 2009, so it's possible many XML readers already parse Unicode names just fine (I've not tested this though and am sure you have more experience with these tools).

One thought for people that would convert from PDML to XML, if the XML tools they're using do not have this support maybe the current PDML-to-XML converter or an additional tool could run through the names in a PDML document and let the user know if any would not work for those XML tools.

The XML specification does mention some prohibited characters, some of which might also make sense to prohibit in PDML. Anyway, more reading and thinking to do there..

While reading the XML specification I was trying to look up each individual Unicode character mentioned, but then I found a [site that categorizes them](https://jrgraphix.net/r/Unicode/) by range, then inside that range you can view all its characters; pretty neat.

**Comment by pml-lang** [comment]
Thanks a lot for the links. Very useful.

I deliberately wanted to start with a rather restrictive rule for names. It will always be easy to loosen the rule, but making it more restrictive in a future version would lead to breaking changes, and must therefore be avoided.

> many people across the world don't speak or write English, or would prefer to use their native language when possible.

I totally agree.

A first step should be to allow all _Unicode letters_, not just [Basic Latin](https://jrgraphix.net/r/Unicode/0020-007F) letters. This will allow people to use their native language, e.g.

```
[quality 10/10]    // English
[Qualität 10/10]   // German
[คุณภาพ 10/10]     // Thai

```

> Starting a "name", or key rather, with a digit can be useful in several cases.

Yes, but it would make PDML names incompatible with XML names (because XML names are not allowed to start with a digit).

Hence, an invalid node like this:

```
[2021
    [- Data -]
]
```
... must currently be written as follows (for example):

```
[Y2021
    [- Data -]
]
```

> allowing the name to be empty

See chapter 'Anonymous Nodes' of our previous [discussion](https://github.com/pdml-lang/pdml-lang.github.io/discussions/15#discussioncomment-5073878), where we agreed that `_` could be used for name-less nodes (e.g. `[_ foo]`).

> names should be case-sensitive.

Yes.

> Reserved "Names" ... For example, one day there might be a desire to use '!' to mean something, or '@', etc.

Yes, `-` and `@` are used already, and more symbols might be used in the future.

Note: Single-letter namespaces are also reserved (this will later be mentioned in chapter 'Namespaces' of the [PDML Extensions User Manual](https://pdml-lang.dev/docs/extensions/user_manual/index.html)).
Currently, only `s`, `u`, and `t` are predefined PDML namespaces, but more namespaces will propably be defined in the future.

> maybe the current PDML-to-XML converter or an additional tool could run through the names in a PDML document and let the user know if any would not work for those XML tools.

Yes.

Converting PDML to XML is currently ok, because PDML uses a subset of the characters allowed in XML names.
However, converting XML to PDML can result in invalid PDML names. Instead of emitting an error, the converter could alternatively replace invalid characters with a valid character explicitly specified on the command line. However, there should not be a _default_ replacement, because that could lead to nasty conversion errors.

> more reading and thinking to do

Yes.

It's better not to rush, and loosen the rules progressively, and only if there are convincing arguments.

**Reply by slanden** [reply]
> It will always be easy to loosen the rule, but making it more restrictive in a future version would lead to breaking changes, and must therefore be avoided

True.

> A first step should be to allow all Unicode letters

Maybe not all; If your goal is to be backward- and forward-compatible with XML, names could match XML's naming rules

```
StartChar: [A-Z] | "_" | [a-z]
           | [#xC0-#xD6] 
           | [#xD8-#xF6]
           | [#xF8-#x2FF]
           | [#x370-#x37D]
           | [#x37F-#x1FFF]
           | [#x200C-#x200D]
           | [#x2070-#x218F]
           | [#x2C00-#x2FEF]
           | [#x3001-#xD7FF]
           | [#xF900-#xFDCF]
           | [#xFDF0-#xFFFD]
           | [#x10000-#xEFFFF]

NameChar: StartChar | "-" | "." | [0-9] | #xB7 | [#x0300-#x036F] | [#x203F-#x2040]
```

The only difference here is that XML additionally allows ':' as a starting character, but I'm not sure why.. Empty namespace?

The above range also already excludes common symbols you might want to use in the future for special meaning.

> because XML names are not allowed to start with a digit

Good point.

> we agreed that _ could be used for name-less nodes

Yes, I like that.

**Reply by pml-lang** [reply]
> Maybe not all; If your goal is to be backward- and forward-compatible with XML, names could match XML's naming rules

Yes, that's what I meant by Unicode _letters_, but missed to explain.
Thanks for clarifying.

**Comment by pml-lang** [comment]
The initial version of PDML was designed to maintain compatibility with XML naming conventions, but this requirement has since been removed.

You presented strong arguments supporting the relaxation of rules for PDML node names.

The great news is that the upcoming PDML version will support all Unicode code points in names, with the exception of control characters. Reserved characters, such as [ and ], must be escaped in names (in the same way they must be escaped in text, e.g. \\[ and \\]).

Everything has already been implemented, but the documentation still needs to be updated. The new version is expected to be released in the first quarter of next year. The update will include several additional improvements, along with some backward-incompatible changes to PDML extensions, as well as new features.

**Comment by pdml-lang** [comment]
The latest published PDML version supports all Unicode code points in node tags/names.
See the [Core PDML 2.0.0 Released](https://github.com/pdml-lang/pdml-lang.github.io/discussions/20) announcement for more information.

@slanden: Thanks for starting this discussion.
